core/src/main/scala/sttp/tapir/server/interceptor/metrics/MetricsEndpointInterceptor.scala

```diff
@@ -0,0 +1,115 @@
+package sttp.tapir.server.interceptor.metrics
+
+import sttp.monad.MonadError
+import sttp.monad.syntax._
+import sttp.tapir.metrics.{EndpointMetric, Metric}
+import sttp.tapir.model.{ServerRequest, ServerResponse}
+import sttp.tapir.server.interceptor._
+import sttp.tapir.server.interpreter.BodyListener
+import sttp.tapir.server.interpreter.BodyListenerSyntax._
+import sttp.tapir.{DecodeResult, Endpoint}
+
+import scala.util.{Success, Try}
+
+class MetricsRequestInterceptor[F[_], B](metrics: List[Metric[F, _]], ignoreEndpoints: Seq[Endpoint[_, _, _, _]])
+    extends RequestInterceptor[F, B] {
+
+  override def apply(responder: Responder[F, B], requestHandler: EndpointInterceptor[F, B] => RequestHandler[F, B]): RequestHandler[F, B] =
+    new RequestHandler[F, B] {
+      override def apply(request: ServerRequest)(implicit monad: MonadError[F]): F[Option[ServerResponse[B]]] =
+        metrics
+          .foldLeft(monad.unit(List.empty[EndpointMetric[F]])) { (mAcc, metric) =>
+            for {
+              metrics <- mAcc
+              endpointMetric <- metric match {
+                case Metric(m, onRequest) => onRequest(request, m, monad)
+              }
+            } yield metrics :+ endpointMetric
+          }
+          .flatMap { endpointMetrics =>
+            requestHandler(new MetricsEndpointInterceptor[F, B](endpointMetrics, ignoreEndpoints)).apply(request)
+          }
+    }
+}
+
+private[metrics] class MetricsEndpointInterceptor[F[_], B](
+    endpointMetrics: List[EndpointMetric[F]],
+    ignoreEndpoints: Seq[Endpoint[_, _, _, _]]
+) extends EndpointInterceptor[F, B] {
+
+  override def apply(responder: Responder[F, B], endpointHandler: EndpointHandler[F, B]): EndpointHandler[F, B] =
+    new EndpointHandler[F, B] {
+      override def onDecodeSuccess[I](
+          ctx: DecodeSuccessContext[F, I]
+      )(implicit monad: MonadError[F], bodyListener: BodyListener[F, B]): F[ServerResponse[B]] = {
+        if (ignoreEndpoints.contains(ctx.endpoint)) endpointHandler.onDecodeSuccess(ctx)
+        else {
+          val responseWithMetrics: F[ServerResponse[B]] = for {
+            _ <- collectMetrics { case EndpointMetric(Some(onRequest), _, _) => onRequest(ctx.endpoint) }
+            response <- endpointHandler.onDecodeSuccess(ctx)
+            withMetrics <- withBodyOnComplete(response) {
+              case Success(_) => collectMetrics { case EndpointMetric(_, Some(onResponse), _) => onResponse(ctx.endpoint, response) }
+              case _          => ().unit
+            }
+          } yield withMetrics
+
+          responseWithMetrics.handleError { case e: Exception =>
```

Comment:
`onException` metrics are collected here, so in `withBodyOnComplete` `case Failure` is left empty, since it could collect `onException` twice i think. If body is a stream which fails before the response is returned, then it will be collected in body listener and here in `handleError`?

Reply:
I think both should be handled, as these are two different scenarios:
1. the effect creating the `ServerResponse` completes successfully, yielding an instance of `ServerResponse` with a streaming body. However, the streaming body might fail when streamed to the client.
2. the above effect fails, so we never even get a `ServerResponse`

Note that a request can be considered "complete" only after 1. successfully finishes (the stream is done). But it can fail either with 1. or 2., but not with both.